Add documentation for Iceberg support in PrestoCPP (#24741)
prestodb/presto: presto-docs/src/main/sphinx/connector/iceberg.rst

    hive.s3.endpoint=http://192.168.0.103:9878
    hive.s3.path-style-access=true

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> ``HIVE``, ``NESSIE``, ``REST``, and ``HADOOP`` Iceberg catalogs are supported in PrestoCPP.

Reviewer: ```suggestion
``HIVE``, ``NESSIE``, ``REST``, and ``HADOOP`` Iceberg catalogs are supported in Presto C++.
```

The following configuration properties are available for all catalog types:

======================================================= ============================================================= ================================== =================== =========================================
Property Name                                           Description                                                   Default                            Presto Java Support Presto C++ Support
======================================================= ============================================================= ================================== =================== =========================================
``iceberg.catalog.type``                                The catalog type for Iceberg tables. The available values     ``HIVE``                           Yes                 Yes, only needed on coordinator
                                                        are ``HIVE``, ``HADOOP``, and ``NESSIE`` and ``REST``.

> ``iceberg.hadoop.config.resources``                     The path(s) for Hadoop configuration resources.                                                  Yes

Reviewer 2: The Presto C++ Support value is missing

                                                        Example: ``/etc/hadoop/conf/core-site.xml.`` This property
                                                        is required if the iceberg.catalog.type is ``hadoop``.
                                                        Otherwise, it will be ignored.

> ``iceberg.file-format``                                 The storage file format for Iceberg tables. The available     ``PARQUET``                        Yes                 NA

Reviewer 2: Can we add a little explanation here why it's NA? We can say "NA, write is not supported yet"

Reviewer 2: Same for other NA items. Thanks!

``iceberg.file-format``                                 The storage file format for Iceberg tables. The available     ``PARQUET``
``iceberg.file-format``                                 The storage file format for Iceberg tables. The available     ``PARQUET``                        Yes                 NA
                                                        values are ``PARQUET`` and ``ORC``.

``iceberg.compression-codec``                           The compression codec to use when writing files. The          ``GZIP``
``iceberg.compression-codec``                           The compression codec to use when writing files. The          ``GZIP``                           Yes                 NA
                                                        available values are ``NONE``, ``SNAPPY``, ``GZIP``,
                                                        ``LZ4``, and ``ZSTD``.

``iceberg.max-partitions-per-writer``                   The Maximum number of partitions handled per writer.          ``100``
``iceberg.max-partitions-per-writer``                   The Maximum number of partitions handled per writer.          ``100``                            Yes                 NA

> ``iceberg.minimum-assigned-split-weight``               A decimal value in the range (0, 1] is used as a minimum      ``0.05``

Author: I think this is because 1 is inclusive but 0 is not for this config. So it would be better to keep it exisitng way

> ``iceberg.delete-as-join-rewrite-enabled``              When enabled, equality delete row filtering is applied        ``true``                           Yes                 Yes

Reviewer 2: I don't think Presto C++ supports it. Presto C++ do not interpret the equality delete file as joins, but would compile it into domain filters or filter functions. Also, reading equality delete file PR is not merged yet. Just say No for now.

Author: Sure, I missed equality delete here.

                                                        as a join with the data of the equality delete files.

> ``iceberg.enable-parquet-dereference-pushdown``         Enable parquet dereference pushdown.                          ``true``                           Yes

Reviewer 2: The Presto C++ Support value is missing

> ``iceberg.enable-parquet-dereference-pushdown``         Enable parquet dereference pushdown.                          ``true``                           Yes                 NA

Reviewer 2: Why is it NA? If it's fails now, just say No

``iceberg.statistic-snapshot-record-difference-weight`` The amount that the difference in total record count matters                                     Yes                 Yes, only needed on coordinator
                                                        when calculating the closest snapshot when picking
                                                        statistics. A value of 1 means a single record is equivalent
                                                        to 1 millisecond of time difference.

``iceberg.pushdown-filter-enabled``                     Experimental: Enable filter pushdown for Iceberg. This is     ``false``                          No                  Yes

>                                                         only supported with Native Worker.

Reviewer: ```suggestion
                                                        only supported with Presto C++.
```
Is this suggestion correct?

> ``iceberg.rows-for-metadata-optimization-threshold``    The maximum number of partitions in an Iceberg table to       ``1000``                           Yes                 Yes

Reviewer 2: Is this one on coordinator only?

Author: As I checked in code its usage is in [IcebergMetadataOptimizer](https://github.com/prestodb/presto/blob/master/presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergMetadataOptimizer.java#L230) so I am assuming this is coordinator only. Do you think, I should check anything more on this?

                                                        metadata log.

``iceberg.metadata-delete-after-commit``                Set to ``true`` to delete the oldest metadata files after     ``false``                          Yes                 NA
                                                        each commit.

``iceberg.metrics-max-inferred-column``                 The maximum number of columns for which metrics               ``100``                            Yes                 NA
                                                        are collected.
``iceberg.max-statistics-file-cache-size``              Maximum size in bytes that should be consumed by the          ``256MB``                          Yes                 Yes, only needed on coordinator
                                                        statistics file cache.
======================================================= ============================================================= ================================== =================== =========================================

> Table Properties

Reviewer 2: Can you please also add a section for the C++ support for Table Properties? Just saying it's not suported because write is not implemented yet

``commit_retries``                        ``commit.retry.num-retries``
``delete_mode``                           ``write.delete.mode``
``metadata_previous_versions_max``        ``write.metadata.previous-versions-max``
``metadata_delete_after_commit``          ``write.metadata.delete-after-commit.enabled``
``metrics_max_inferred_column``           ``write.metadata.metrics.max-inferred-column-defaults``
=======================================   =======================================================

Session Properties
------------------

Session properties set behavior changes for queries executed within the given session.

===================================================== ======================================================================= =================== =============================================
Property Name                                         Description                                                             Presto Java Support Presto C++ Support
===================================================== ======================================================================= =================== =============================================

> ``iceberg.delete_as_join_rewrite_enabled``            Overrides the behavior of the connector property                        Yes                 No, equality delete read is not supported yet

Reviewer: ```suggestion
``iceberg.delete_as_join_rewrite_enabled``            Overrides the behavior of the connector property                        Yes                 Equality delete read is not supported
```

Author: updated

``metrics_max_inferred_column``           ``write.metadata.metrics.max-inferred-column-defaults``
=======================================   ===============================================================

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> Table properties are not supported in Presto C++ because write operations have not been implemented.

Reviewer 3: I think we should consider re-wording this. It's not quite clear what the statement "Table properties are not supported ..." means. Does that mean reading them? writing them? considering them during writes? I think we should be a bit more specific

Reviewer 2: > I think we should consider re-wording this. It's not quite clear what the statement "Table properties are not supported ..." means. Does that mean reading them? writing them? considering them during writes? I think we should be a bit more specific

I think what @ZacBlanco said makes sense. @agrawalreetika Will you be able to try them and update the doc separating the read and write

Session Properties
------------------

Session properties set behavior changes for queries executed within the given session.

===================================================== ======================================================================= =================== ==================
Property Name                                         Description                                                             Presto Java Support Presto C++ Support
===================================================== ======================================================================= =================== ==================

> ``iceberg.delete_as_join_rewrite_enabled``            Overrides the behavior of the connector property                        Yes                 Yes

Reviewer 2: Should be No, or NA

                                                       this size will not be cached.
====================================================   =============================================================   ============

Presto C++ Support
~~~~~~~~~~~~~~~~~~

> Manifest file caching is supported in PrestoCPP.

Reviewer: ```suggestion
Manifest file caching is supported in Presto C++.
```

    SHOW TABLES FROM jmx.current like '%alluxio%';

Presto C++ Support
~~~~~~~~~~~~~~~~~~

> Alluxio data caching is applicable for Presto Java. PrestoCPP supports Async data cache. See :ref:`async_data_caching_and_prefetching` for details.

Reviewer: ```suggestion
Alluxio data caching is applicable for Presto Java. Presto C++ supports Async data cache. See :ref:`async_data_caching_and_prefetching` for details.
```

    SELECT * FROM jmx.current."com.facebook.presto.hive:name=iceberg_parquetmetadata,type=cachestatsmbean";

Presto C++ Support
~~~~~~~~~~~~~~~~~~

> File and stripe footer cache is not applicable for PrestoCPP.

Reviewer: ```suggestion
File and stripe footer cache is not applicable for Presto C++.
```

     ----------------------------------+------------
                  2                    | 3

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> All above extra hidden metadata columns are supported in PrestoCPP.

Reviewer: ```suggestion
All :ref:`connector/iceberg:Extra Hidden Metadata Columns` are supported in Presto C++.
```
Tested link in local doc build.

     ----------------------------------+------------
                  2                    | 3

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> All above extra hidden metadata columns are supported in Presto C++.

Reviewer 3: ```suggestion
All above metadata columns are supported in Presto C++.
```

> above extra hidden

Three adjectives in a row feels like a lot. I think we can reduce

Author: Here existing `extra hidden metadata columns` is taken from the main heading. But if we think we should just rephrase it saying - `All above metadata columns are supported in Presto C++` that's fine with me?

Reviewer 3: It feel overly description to have three adjectives in the sentence - I think this gets the point across

      testBranch | BRANCH | 3374797416068698476 | NULL                    | NULL                  | NULL
      testTag    | TAG    | 4686954189838128572 | 10                      | NULL                  | NULL

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> All above extra hidden metadata tables are supported in PrestoCPP.

Reviewer: ```suggestion
All :ref:`connector/iceberg:Extra Hidden Metadata Tables` are supported in Presto C++.
```

      testBranch | BRANCH | 3374797416068698476 | NULL                    | NULL                  | NULL
      testTag    | TAG    | 4686954189838128572 | 10                      | NULL                  | NULL

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> All above extra hidden metadata tables are supported in Presto C++.

Reviewer 2: Just want to double confirm that you have verified they are working, right?

Author: Apart from`$changelog` everything works, `$changelog` throws an error which I missed adding here. I opened an issue for the same. https://github.com/prestodb/presto/issues/24816

> All above extra hidden metadata tables, except `$changelog`, are supported in Presto C++.

Reviewer 3: ```suggestion
All above metadata tables, except `$changelog`, are supported in Presto C++.
```

Author: Here existing `extra hidden metadata tables` is taken from the main heading. But if we think we should just rephrase it saying - `All above metadata tables, except `$changelog`, are supported in Presto C++.` that's fine with me, LMK.

    CALL iceberg.system.set_table_property('schema_name', 'table_name', 'commit.retry.num-retries', '10');

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> All above procedures are supported in PrestoCPP.

Reviewer: ```suggestion
All :ref:`connector/iceberg:Procedures` are supported in Presto C++.
```

SQL Support
-----------

SQL Support Summary for Presto Java and Presto CPP:

============================== ============= ============ ============================================================================

> SQL Operation                  Presto Java   Presto CPP   Comments

Reviewer: ```suggestion
SQL Operation                  Presto Java   Presto C++   Comments
```

============================== ============= ============ ============================================================================

> ``CREATE SCHEMA``              ✅            ✅

Reviewer: Consider changing the ✅ and ❌ in this table to the Yes and No you used in Configuration Properties and Session Properties.

``CREATE TABLE``               ✅            ✅

``CREATE VIEW``                ✅            ✅

``INSERT INTO``                ✅            ❌

``CREATE TABLE AS SELECT``     ✅            ❌

> ``SELECT``                     ✅            ✅             Read is supported in PrestoCPP including those with positional delete files.

Reviewer: ```suggestion
``SELECT``                     ✅            ✅             Read is supported in Presto C++ including those with positional delete files.
```

``Hour``              ``timestamp``
===================== =======================================================================

Presto C++ Support
~~~~~~~~~~~~~~~~~~

> Read from the tables with Partition column transform is supported in Presto C++.

Reviewer 3: ```suggestion
Reads from tables with partition column transforms is supported in Presto C++.
```

evolution, users can evolve a table schema with SQL after enabling the Presto
Iceberg connector.

> Presto C++ Support

Reviewer 2: This is for Partition Column Transform. Could you please add Presto C++ support as well? Thanks!

^^^^^^^^^^^^^^^^^^

> Schema Evolution is supported in PrestoCPP.

Reviewer: ```suggestion
:ref:`connector/iceberg:Schema Evolution` is supported in Presto C++.
```

Parquet Writer Version
----------------------

Presto now supports Parquet writer versions V1 and V2 for the Iceberg catalog.
It can be toggled using the session property ``parquet_writer_version`` and the config property ``hive.parquet.writer.version``.
Valid values for these properties are ``PARQUET_1_0`` and ``PARQUET_2_0``. Default is ``PARQUET_1_0``.

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> PrestoCPP supports Parquet writer versions V1.

Reviewer: ```suggestion
Presto C++ supports Parquet writer versions V1.
```

evolution, users can evolve a table schema with SQL after enabling the Presto
Iceberg connector.

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> Schema Evolution is supported in Presto C++.

Reviewer 3: ```suggestion
Schema evolution is supported in Presto C++.
```

Parquet Writer Version
----------------------

Presto now supports Parquet writer versions V1 and V2 for the Iceberg catalog.
It can be toggled using the session property ``parquet_writer_version`` and the config property ``hive.parquet.writer.version``.
Valid values for these properties are ``PARQUET_1_0`` and ``PARQUET_2_0``. Default is ``PARQUET_1_0``.

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> Presto C++ supports Parquet writer versions V1.

Reviewer 3: This is a little confusing because earlier the properties section stated that writes are not supported?

Author: Thanks for pointing it point.
It's about actually reading, so I think I can rephrase it something like - `Presto C++ supports reading Parquet data written with Parquet writer version V1.` ?

evolution, users can evolve a table schema with SQL after enabling the Presto
Iceberg connector.

Presto C++ Support
^^^^^^^^^^^^^^^^^^

Schema evolution is supported in Presto C++.

Parquet Writer Version
----------------------

Presto now supports Parquet writer versions V1 and V2 for the Iceberg catalog.
It can be toggled using the session property ``parquet_writer_version`` and the config property ``hive.parquet.writer.version``.
Valid values for these properties are ``PARQUET_1_0`` and ``PARQUET_2_0``. Default is ``PARQUET_1_0``.

Presto C++ Support
^^^^^^^^^^^^^^^^^^

Presto C++ supports reading Parquet data written with Parquet writer version V1.

> 

Reviewer 3: I see that across these `Presto C++ Support` sections we have varying forms

> ... is supported in Presto C++

and

> Presto C++ supports ....

Two comments on this

1. We should be consistent
2. I prefer that we standardize on the 2nd as it's in active voice, which is generally the form we want to write for our documentation (@steveburnett can you confirm the 2nd is the form that we should prefer?)

Reviewer: Yes, we should prefer the use of active voice, but exceptions where passive voice is better can happen. Sometimes active voice can be more awkward depending on the emphasis you want to communicate. 

See 
* the Gitlab Documentation Style Guide entry for [Active Voice](https://docs.gitlab.com/development/documentation/styleguide/#active-voice) (shorter)
* the Google Developer Documentation Style Guide entry for [Active Voice](https://developers.google.com/style/voice) (longer and more examples)

tl;dr: use active voice, but if active voice is awkward then passive voice is okay.

Reviewer: In this context of the "Presto C++ Support" subtopics, I would argue that passive voice is better. 

* The statements immediately follow the heading "Presto C++ Support" so "Presto C++ supports" repeat the information without adding to it, and 

* Passive voice here follows the guideline in the Google Developer Documentation Style Guide entry for [Active Voice](https://developers.google.com/style/voice) "To emphasize an object over an action." 

Using passive voice here emphasizes the object "HIVE, NESSIE, REST, and HADOOP Iceberg catalogs" and makes it easier for the user to find the information they need. 

See the screenshots I made in a local doc build to show the difference. I think the second one is the better one. 

1)
<img width="610" alt="Screenshot 2025-04-08 at 10 20 42 AM" src="https://github.com/user-attachments/assets/7d05aa19-06a0-4de6-8a65-ca10c839b56a" />

2)
<img width="688" alt="Screenshot 2025-04-08 at 10 20 15 AM" src="https://github.com/user-attachments/assets/49d0aac8-cd20-4406-86ae-2acce678a090" />

            20 | canada        |         2 | comment
    (3 rows)

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> Time Travel is supported in PrestoCPP.

Reviewer: ```suggestion
Time Travel is supported in Presto C++.
```

            20 | canada        |         2 | comment
    (3 rows)

Presto C++ Support
^^^^^^^^^^^^^^^^^^

> Time Travel is supported in Presto C++.

Reviewer 3: ```suggestion
Time travel queries are supported in Presto C++.
```
